motionplan/motionPlanner.go

```diff
@@ -488,80 +487,71 @@ func CheckPlan(
 	// The solver frame will have had its PTGs filled in the newPlanManager() call, if applicable.
 	relative := len(sf.PTGSolvers()) > 0
 
-	if relative {
-		// get pose of robot along the current trajectory it is executing
-		lastPose, err := sf.Transform(currentInputs)
-		if err != nil {
-			return err
-		}
-
-		// where ought the robot be on the plan
-		pathPosition := spatialmath.PoseBetweenInverse(errorState, currentPosition)
-
-		// absolute pose of the previous node we've passed
-		formerRunningPose := spatialmath.PoseBetweenInverse(lastPose, pathPosition)
-
-		// convert planNode's poses to be in absolute coordinates
-		if planNodes, err = rectifyTPspacePath(planNodes, sf, formerRunningPose); err != nil {
-			return err
-		}
+	// offset the plan using the errorState
+	offsetPlan := OffsetPlan(plan, errorState)
+	poses, err := offsetPlan.Path().GetFramePoses(checkFrame.Name())
+	if err != nil {
+		return err
 	}
 
-	// adjust planNodes by the errorState
-	// this only changes a node's pose and not its inputs
-	planNodes = transformNodes(planNodes, errorState)
-
-	// pre-pend node with current position of robot to planNodes
-	// Note that currentPosition is assumed to have accounted for the errorState
-	// Note that currentInputs is assumed to have NOT accounted for the errorState
-	planNodes = append([]node{&basicNode{pose: currentPosition, q: currentInputs}}, planNodes...)
-
-	// create constraints
+	// setup the planOpts
 	if sfPlanner.planOpts, err = sfPlanner.plannerSetupFromMoveRequest(
-		currentPosition,                    // starting pose
-		planNodes[len(planNodes)-1].Pose(), // goalPose
-		plan[0],                            // starting configuration
+		currentPose,
+		poses[len(poses)-1],
+		currentInputs,
 		worldState,
 		nil, // no pb.Constraints
 		nil, // no plannOpts
 	); err != nil {
 		return err
 	}
 
-	// go through plan and check that we can move from plan[i] to plan[i+1]
-	var totalTravelDistanceMM float64
-	for i := 0; i < len(planNodes)-1; i++ {
-		currentPose := planNodes[i].Pose()
-		nextPose := planNodes[i+1].Pose()
-		startConfiguration := planNodes[i].Q()
-		endConfiguration := planNodes[i+1].Q()
-
+	createSegment := func(
+		currPose, nextPose spatialmath.Pose,
+		currInput, nextInput map[string][]frame.Input,
+	) (*ik.Segment, error) {
+		currInputSlice, err := sf.mapToSlice(currInput)
+		if err != nil {
+			return nil, err
+		}
+		nextInputSlice, err := sf.mapToSlice(nextInput)
+		if err != nil {
+			return nil, err
+		}
 		// If we are working with a PTG plan we redefine the startConfiguration in terms of the endConfiguration.
 		// This allows us the properly interpolate along the same arc family and sub-arc within that family.
 		if relative {
-			startConfiguration = []frame.Input{
-				{Value: endConfiguration[0].Value}, {Value: endConfiguration[1].Value}, {Value: 0},
-			}
+			currInputSlice = []frame.Input{{Value: nextInputSlice[0].Value}, {Value: nextInputSlice[1].Value}, {Value: 0}}
```

Comment:
I think i just found another 🐛 ! I do not think the bug exists on main, but instead exists as an incompleteness of the function..

---------------------

We have our Arm + Gantry test `TestArmGantryCheckPlan`  which does not does not use TPspace. Hence, within this function `relative` is false and we never over-write `currInputSlice`. 

But what about an arm mounted on a base (using PTG kinematics)?
In this case, the `solverFrame` we construct would have `3 + 6 DOF` and since we are relative we would over-write it back to 3 DOF. 

I think, for the time being, the `relative` conditional should remain, but future work should make this conditional smarter..


---------------------


I do not think it's worth to handle in this PR since this functionality (arm + base) planning is still experimental, but still worth noting through ticket or comment or something else imo ..

Reply:
Yeah the behavior of base + arm is undefined and unsupported.  It also appears to me that this same problem would exist in main

Reply:
> behavior of base + arm is undefined

It's not undefined; it's explicitly defined as "throw an error" :)

Reply:
Even better :)

Reply:
The solver frame will refuse to initialize if you try to plan for a base + arm simultaneously.
https://github.com/viamrobotics/rdk/blob/a22ad5bd00c60d3277c380740d1b7e39e7f5be82/motionplan/solverFrame.go#L164